### PR TITLE
Making JsonFactory and HttpFactory Configurable

### PR DESCRIPTION
**Pull request by hiranya911**
* Exposed protected factory methods in `FirebaseCredentials`, which accept `JsonFactory` and `HttpTransport` arguments.
* Added `setJsonFactory()` and `setHttpTransport()` methods to `FirebaseOptions`.

Related to #35

### CODE REVIEW

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 1]
```diff
@@ -71,6 +76,26 @@ FirebaseCredential getCredential() {
...
+   * @return A Google API client JsonFactory instance.
+   */
+  @NonNull
+  public JsonFactory getJsonFactory() {
+    return jsonFactory;
+  }
+
   @Override
   public boolean equals(Object obj) {
     if (!(obj instanceof FirebaseOptions)) {
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: Should equals / hashCode / toString need to be updated?

File: src/main/java/com/google/firebase/auth/FirebaseCredentials.java [hunk 2]
```diff
@@ -80,8 +79,26 @@ public static FirebaseCredential applicationDefault() {
...
+   *
+   * <p>See <a
+   * href="https://developers.google.com/identity/protocols/application-default-credentials">Google
+   * Application Default Credentials</a> for details on Google Application Deafult Credentials.
+   *
+   * <p>See <a href="/docs/admin/setup#initialize_the_sdk">Initialize the SDK</a> for code samples
+   * and detailed documentation.
+   *
+   * @param transport HttpTransport used to communicate with the remote authentication server.
+   * @param jsonFactory JsonFactory used to parse JSON responses from the remote authentication
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/FirebaseCredentials.java]: from the remote authentication ....   server?

File: src/main/java/com/google/firebase/FirebaseOptions.java (hunk 1 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: It looks like these were implemented to support some persistence requirements, which are not really used in the SDK anymore. In the future I can see us adding even more options to this class, and having to implement all these methods each time we make a change is a bit of a problem. WDYT? Should we just get rid of the implementations?

`FirebaseApp` also implements these methods, but at least it performs all comparisons based on the name of the app.

File: src/main/java/com/google/firebase/auth/FirebaseCredentials.java (hunk 2 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseCredentials.java]: Done

**Comment by hiranya911** [comment]
Fixed the typo, and left a follow up question.

File: src/main/java/com/google/firebase/FirebaseOptions.java (hunk 1 above)
Comment (mikelehen) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: Deleting it sounds good to me.

Comment (hiranya911) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: I will do this in a separate PR. The existing implementations in `FirebaseOptions` are ok for now.

Comment (mikelehen) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: Please do this sooner than later.  Leaving broken equals / hashCode code around is a good recipe for unhappiness.

**Pull request merged**